NDTensors/ext/NDTensorCUDA/set_types.jl

```diff
@@ -0,0 +1,34 @@
+buffertype(datatype::Type{<:CuArray{<:Any,<:Any,B}}) where {B} = B
+function buffertype(datatype::Type{<:CuArray})
+  println(
+    "CuArray definitions require a CUDA.Mem buffer try $(datatype{default_buffertype()})"
+  )
+  throw(TypeError)
+end
+
+default_buffertype() = CUDA.Mem.DeviceBuffer
+
+function set_eltype(arraytype::Type{<:CuArray}, eltype::Type)
+  return CuArray{eltype,NDTensors.ndims(arraytype),buffertype(arraytype)}
+end
+
+function set_ndims(arraytype::Type{<:CuArray{T,<:Any,<:Any}}, ndims) where {T}
+  return CuArray{eltype(arraytype),ndims,buffertype(arraytype)}
+end
+
+function set_ndims(arraytype::Type{<:CuArray{T}}, ndims) where {T}
+  return CuArray{eltype(arraytype),ndims,default_buffertype()}
```

Comment:
It seems too opinionated to fill in the `default_buffertype()` here, shouldn't it leave it unspecified if it wasn't specified?

In a future PR we should start using `SetParameters`, it should make these types of functions simpler to define.